api/resolver.py

```diff
@@ -1,177 +1,361 @@
-        'subject.info':      0,
-        'subject.firstname': 0,
-        'subject.lastname':  0,
+    # In some cases we only want to resolve the id of a container
+    id_only_projection = {
+        'label':             1,
+        'permissions':       1,
+        'files':             1,
     }
 
-    # Add some more fields for debugging purposes.
-    # projection['roles']       = 0
-    # projection['permissions'] = 0
+    def __init__(self, cont_name, files=True, use_id=False, analyses=True):
+        self.cont_name = cont_name
+        self.storage = ContainerStorage.factory(cont_name)
+        # container_type is also the parent id field name
+        self.container_type = containerutil.singularize(cont_name)
+        self.files = files
+        self.use_id = use_id        
+        self.analyses = analyses
+        self.child_name = self.storage.get_child_container_name()
+
+    def next(self, path_in, path_out, id_only):
+        """Get the next node in the hierarchy, adding any value found to path_out"""
+        # If there is no path in, don't try to resolve
+        if not path_in:
+            return None
+
+        use_id, criterion = parse_criterion(path_in)
+        parent = get_parent(path_out)
+        # Peek to see if we need files for the next path element
+        fetch_files = (path_peek(path_in) in ['files', None])
+
+        # Setup criterion match
+        query = {}
+        if criterion:
+            if use_id or self.use_id:
+                if self.storage.use_object_id:
+                    try:
+                        query['_id'] = bson.ObjectId(criterion)
+                    except bson.errors.InvalidId as e:
+                        raise InputValidationException(e.message)
+                else:
+                    query['_id'] = criterion
+            else:
+                query['label'] = criterion
+
+        # Setup projection
+        if id_only: 
+            proj = ContainerNode.id_only_projection.copy()
+            if fetch_files:
+                proj['files'] = 1
+        else:
+            proj = self.storage.get_list_projection()
+            if proj and not fetch_files:
+                proj['files'] = 0
+
+        # We don't use the user field here because we want to return a 403 if
+        # they try to resolve something they don't have access to
+        results = self.find(query, parent, proj)
+        if not results:
+            raise APINotFoundException('No {0} {1} found.'.format(criterion, self.container_type))
+        
+        child = results[0]
+
+        self.storage.filter_deleted_files(child)
+        child['container_type'] = self.container_type
+        path_out.append(child)
+
+        # Get the next node
+        if not path_in:
+            return None
+
+        # Files
+        if fetch_files:
+            path_in.popleft()
+            return FilesNode()
+
+        # Check for analyses
+        if path_peek(path_in) == 'analyses':
+            if self.analyses:
```

Comment:
I think these two conditionals should be combined and the `raise` removed. Right now, there's nothing stopping someone calling a container `analyses` (ref https://github.com/scitran/core/pull/1089#issuecomment-370565314) and I would expect a not-found error to occur if nothing is found.